Source/WebGPU/WebGPU/Surface.mm

```diff
@@ -29,9 +29,34 @@
 #import "APIConversions.h"
 #import "Adapter.h"
 
+#import <WebGPU/WebGPUExt.h>
+
+static IOSurfaceRef getSurfaceFromDescriptor(const WGPUSurfaceDescriptor& descriptor)
+{
+    return (descriptor.nextInChain && descriptor.nextInChain->sType == static_cast<WGPUSType>(WGPUSTypeExtended_SurfaceDescriptorCocoaSurfaceBacking)) ? ((const WGPUSurfaceDescriptorFromCocoaCustomSurface*)descriptor.nextInChain)->surface : nullptr;
+}
+
+IOSurfaceRef wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer(WGPUSurface surface)
+{
+    return WebGPU::fromAPI(surface).surface();
+}
+
+WGPUSurface wgpuInstanceCreateSurface(WGPUInstance, const WGPUSurfaceDescriptor* descriptor)
+{
+    return WebGPU::releaseToAPI(WebGPU::Surface::create(*descriptor));
+}
+
 namespace WebGPU {
 
-Surface::Surface() = default;
+Ref<Surface> Device::createSurface(const WGPUSurfaceDescriptor& descriptor)
+{
+    return Surface::create(descriptor);
+}
+
+Surface::Surface(const WGPUSurfaceDescriptor& descriptor)
+    : m_surface(getSurfaceFromDescriptor(descriptor))
```

Comment:
So this one should probably not create the iosurface, rather record the properties.
E.g. `wgpuInstanceCreateSurface(300x200)` creates "empty surface" of size 300x200. It doesn't have the backing store by default (unless it's somehow needed by the api.)
The surface should get its contents from the drawing done by the caller via SwapChain present.

Reply:
Don't we need to create the IOSurface prior to binding the texture as the color attachment for the render pass?

I.e., in this MTLDevice call:
```
- (nullable id <MTLTexture>)newTextureWithDescriptor:(MTLTextureDescriptor *)descriptor iosurface:(IOSurfaceRef)iosurface plane:(NSUInteger)plane API_AVAILABLE(macos(10.11), ios(11.0));
 ```

we pass the IOSurface already created so that when the render pass completes, we can just call `layer.setContents(ioSurface)`.

I could certainly defer creation of the IOSurface to the time `GPUCanvasContext.getCurrentTexture` is called but I'm not sure how to defer the IOSurface creation later than that point